### ISSUE TITLE
Illuminate Schema Blueprint 

### ISSUE THREAD

**Issue by treehousetim** [mstaack/laravel-postgis#166]
Similar to #81
Similar to #109 

I'm getting this issue in 2022.

laravel.ERROR: MyMigration::{closure}():
Argument 1 ($table) must be of type MStaack\LaravelPostgis\Schema\Blueprint,
Illuminate\Database\Schema\Blueprint given,

called in /opt/site/vendor/illuminate/database/Schema/Blueprint.php on line 95

[stacktrace]

  0. /opt/site/vendor/illuminate/database/Schema/Blueprint.php(95): MyMigration->{closure}()
  1. /opt/site/vendor/illuminate/database/Schema/Builder.php(385): Illuminate\\Database\\Schema\\Blueprint->__construct()
  2. /opt/site/vendor/illuminate/database/Schema/Builder.php(212): Illuminate\\Database\\Schema\\Builder->createBlueprint()
  3. /opt/site/vendor/illuminate/support/Facades/Facade.php(261): Illuminate\\Database\\Schema\\Builder->table()
  4. /opt/site/database/migrations/2022_02_24_000000_my_migration.php(26): Illuminate\\Support\\Facades\\Facade::__callStatic()

Line 19 is the first line of this code block
```php
Schema::table( 'my_table', function( Blueprint $table ) {
    $table->geometrycollection( 'geometry' );
});
```
The issue seems to be that the illuminate schema facade calls the illuminate blueprint constructor which calls the callback with `$this` (which is the illuminate Blueprint class itself rather than the MStaack Blueprint class.

I'm uncertain how to fix this.

Here is the use block from the top of my migration file
```<?php

use Illuminate\Database\Migrations\Migration;
use MStaack\LaravelPostgis\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;
```

**Comment by joelmellon** [comment]
`use MStaack\LaravelPostgis\Schema\Blueprint as PostgisBlueprint;` and type-hint your closure param `...function( PostgisBlueprint $table ) {...`

Learn about `inheritance` and `use aliases`, eg. https://www.php.net/manual/en/language.namespaces.importing.php

**Comment by treehousetim** [comment]
Hi @joelmellon  please note:

> The issue seems to be that the illuminate schema facade calls the illuminate blueprint constructor which calls the callback with $this (which is the illuminate Blueprint class itself rather than the MStaack Blueprint class.

Aliases are only used for scope resolution.  Even with your suggestion the error persists and continues to be the exact same error.

```
Argument #1 ($table) must be of type MStaack\LaravelPostgis\Schema\Blueprint, Illuminate\Database\Schema\Blueprint given
```

**Comment by treehousetim** [comment]
See the example from the README of this project below.

My point is that this doesn't and can't work because the Illuminate `Schema` facade is hard coded to send the closure an Illuminate object - it doesn't use reflection with type declarations to determine the type of object to send the closure.

I'm not being an ass, I'm really confused, fairly new to Laravel (which is why I dug into the calling facade class) but not new to PHP. 

```php
use Illuminate\Database\Migrations\Migration;
use MStaack\LaravelPostgis\Schema\Blueprint;

class CreateLocationsTable extends Migration {

    /**
     * Run the migrations.
     *
     * @return void
     */
    public function up()
    {
        Schema::create('locations', function(Blueprint $table)
        {
            $table->increments('id');
            $table->string('name');
            $table->string('address')->unique();
            $table->point('location'); // GEOGRAPHY POINT column with SRID of 4326 (these are the default values).
            $table->point('location2', 'GEOGRAPHY', 4326); // GEOGRAPHY POINT column with SRID of 4326 with optional parameters.
            $table->point('location3', 'GEOMETRY', 27700); // GEOMETRY column with SRID of 27700.
            $table->polygon('polygon'); // GEOGRAPHY POLYGON column with SRID of 4326.
            $table->polygon('polygon2', 'GEOMETRY', 27700); // GEOMETRY POLYGON column with SRID of 27700.
            $table->timestamps();
        });
    }

    /**
     * Reverse the migrations.
     *
     * @return void
     */
    public function down()
    {
        Schema::drop('locations');
    }

}
```

**Comment by joelmellon** [comment]
> it doesn't use reflection with type declarations to determine the type of object to send the closure

Right, so if you remove the Blueprint type-hint from your closure, what happens?

**Comment by treehousetim** [comment]
Thanks for helping me with this.


What version of Laravel are you on?

I'm on Laravel/lumen 8.

From the readme:


> For laravel >=5.5 that's all. This package supports Laravel new [Package Discovery](https://laravel.com/docs/5.5/packages#package-discovery).
>
> If you are using Laravel < 5.5, you also need to add the DatabaseServiceProvider to your config/app.php file.

**Comment by joelmellon** [comment]
Yeah, I'm running Laravel 8.x as well (8.83.2) with PHP 8.1.3, and 5.2.1 of mstaack/laravel-postgis.

You might try putting a dump/die in `vendor/mstaack/laravel-postgis/src/DatabaseServiceProvider.php::boot()` or something to make sure it's getting loaded. https://stackoverflow.com/questions/22444685/extend-blueprint-class

You might also experiment with `blueprintResolver()` to see if you can get that working. It doesn't look this package uses that under the hood, but to be honest, I'm not sure the exact mechanism being used to inject it's own Blueprint.

I'll have some time to dig into that later because I'm getting more and more curious. I'm 100% invested now, lol.

**Comment by joelmellon** [comment]
Ahh yes. The Mstaack provider "overrides" the ConnectionFactory (the 'db.factory' singleton) providing the Mstaack PostgisConnection (when using the 'pgsgl' driver) which provides the Mstaack Schema\Builder which provides the Mstaack Schema\Blueprint.

...so, that means the if the service provider isn't registered, it doesn't work. I'm not sure exactly, but it also seems like this methodology may not be compatible with other third-party Builders or Blueprints. It seems to me like they could conflict. (You may want to check your other packages for any that might offer migration functionality.) In your case though, since you're getting the default Laravel Blueprint, it's more probably that Mstaack just isn't being registered.

My last recommendation is to manually register the Mstacck service provider. There are valid reasons why service providers aren't auto-registered, and it can't really hurt to try it.

> If you are using Laravel < 5.5, you also need to add the DatabaseServiceProvider to your config/app.php file.

...and again, put a `dd()` in the service provider's `register()` method to make sure it's loaded.

**Comment by treehousetim** [comment]
So....

All this time, wasted because auto discovery isn't working (I'm probably ignorant about how to make it work) and when I manually register this service provider I get an error. But according to #43  this won't work with Lumen.  I'm using lumen.

Thanks for all your help, @joelmellon

![Screen Shot 2022-03-01 at 5 42 01 PM](https://user-images.githubusercontent.com/541887/156267482-ab12ea31-0efb-4f7a-8f75-b1e049606c86.png)

**Comment by joelmellon** [comment]
Ahh! I don't think Lumen offers [auto-discovery](https://github.com/laravel/lumen-framework/issues/649) so this whole thing makes perfect sense now. Copy/pasting service provider class names into your `bootstrap/app.php` is trivial though.

Lumen also doesn't have a `config_path()` helper function so that's all the error is. I think in Lumen you'd just do `app()->getConfigurationPath()` if you needed to yourself.

Moving past this issue may end up being a rabbit hole for you, but resolving this piece of the puzzle would probably take 5 minutes.

You can pretty much remove all the code in the boot method of that provider. Obviously you don't want to modify vendor files, so you can just extend it and [register](https://lumen.laravel.com/docs/8.x/providers) your own in [bootstrap/app.php](https://github.com/laravel/lumen/blob/69b26578d2f15595ea901278434b74df459c4329/bootstrap/app.php#L96), eg.

Create `app/Providers/PostgisServiceProvider.php`:
```php
<?php

namespace App\Providers;

use MStaack\LaravelPostgis\DatabaseServiceProvider;

class PostgisServiceProvider extends DatabaseServiceProvider
{
    /**
     * Bootstrap any application services.
     *
     * @return void
     */
    public function boot()
    {
        // do nothing - publishing Mstaack stuff is just conveniently copying files for you
    }

   // the Mstaack provider's register() method will still be called
}

```

Then in `bootstrap/app.php`:
```php
/*
|--------------------------------------------------------------------------
| Register Service Providers
|--------------------------------------------------------------------------
| ...
*/

// $app->register(App\Providers\AppServiceProvider::class);
// $app->register(App\Providers\AuthServiceProvider::class);

// add this 👇
$app->register(App\Providers\PostgisServiceProvider::class);
```

All this does is disable Mstaack's "publishing" capabilities. You can obviously make your own `config/postgis.php` (which only has [one key](https://github.com/mstaack/laravel-postgis/blob/1726acd3f9ca01cb3c3979453aa014ac93bb6c64/config/postgis.php#L4) in it, lol.) Then you can enable the PostGIS extension yourself. Or, just make your own migration that enables it like [the stub](https://github.com/mstaack/laravel-postgis/blob/1726acd3f9ca01cb3c3979453aa014ac93bb6c64/database/migrations/enable_postgis.php.stub#L15) does.

It's quite possible that there's other Laravel specific code in this package, but this "extend and override" principle can be used until it's working for you. It's all starts from the service provider, so from there you have full control over which classes and methods actually end up getting used. First you'd extend and replace the `MStaack\LaravelPostgis\Connectors\ConnectionFactory` class with your own implementation, then the Mstaack Connection class, then as you bump into code that breaks, extend and override that functionality with your own non-Laravel specific code. Depending on how far down the rabbit hole you have to go, it could end up saving you time over writing your own PostGIS functionality.

I'm new to PostGIS (and Postgres) so this package has been quite useful to me, and if I needed it in a Lumen project, I would definitely go down the rabbit hole. I enjoy digging into internals though. ‾\\_(ツ)_/‾

**Comment by mstaack** [comment]
Consider ditching lumen. No real reason to use it anymore.

**Comment by tyekx** [comment]
I ran into this issue having conflicting providers from package `tpetry/laravel-postgresql-enhanced`. I solved it by removing said package, though im sure if u wrote out the entire class name instead of the facade it would probably work?
